src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -0,0 +1,134 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "frozenobjectheap.h"
+#include "memorypool.h"
+
+FrozenObjectHeap::FrozenObjectHeap():
+    m_pStart(nullptr),
+    m_pCurrent(nullptr),
+    m_SizeCommited(0),
+    m_Size(0),
+    m_SegmentHandle(nullptr)
+    COMMA_INDEBUG(m_ObjectsCount(0))
+{
+    m_PageSize = GetOsPageSize();
+    m_Crst.Init(CrstFrozenObjectHeap, CRST_UNSAFE_COOPGC);
+}
+
+#define FOH_RESERVE_PAGES 1024 // e.g. reserve 4Mb of virtual memory
+#define FOH_COMMIT_PAGES 32  // e.g. commit 128Kb chunks
+
+FrozenObjectHeap::~FrozenObjectHeap()
+{
+    if (m_SegmentHandle != nullptr)
+    {
+        GCHeapUtilities::GetGCHeap()->UnregisterFrozenSegment(m_SegmentHandle);
+    }
+
+    if (m_pStart != nullptr)
+    {
+        ClrVirtualFree(m_pStart, 0, MEM_RELEASE);
+    }
+}
+
+bool FrozenObjectHeap::Initialize()
+{
+    m_Size = m_PageSize * FOH_RESERVE_PAGES;
```

Comment:
The OS page size is 64kB on some ARM64 systems. It does not sound right to make this reservation size depend on how the OS is configured.

Reply:
Removed m_PageSize